libsolidity/ast/ASTJsonConverter.cpp

```diff
@@ -270,18 +283,23 @@ bool ASTJsonConverter::visit(ImportDirective const& _node)
 
 bool ASTJsonConverter::visit(ContractDefinition const& _node)
 {
-	setJsonNode(_node, "ContractDefinition", {
+	std::vector<pair<string, Json::Value>> attributes = {
 		make_pair("name", _node.name()),
 		make_pair("documentation", _node.documentation() ? toJson(*_node.documentation()) : Json::nullValue),
 		make_pair("contractKind", contractKind(_node.contractKind())),
 		make_pair("abstract", _node.abstract()),
-		make_pair("fullyImplemented", _node.annotation().unimplementedDeclarations.empty()),
-		make_pair("linearizedBaseContracts", getContainerIds(_node.annotation().linearizedBaseContracts)),
 		make_pair("baseContracts", toJson(_node.baseContracts())),
 		make_pair("contractDependencies", getContainerIds(_node.annotation().contractDependencies, true)),
 		make_pair("nodes", toJson(_node.subNodes())),
 		make_pair("scope", idOrNull(_node.scope()))
-	});
+	};
+
+	if (_node.annotation().unimplementedDeclarations.has_value())
+		attributes.emplace_back("fullyImplemented", _node.annotation().unimplementedDeclarations->empty());
+	if (!_node.annotation().linearizedBaseContracts.empty())
+		attributes.emplace_back("linearizedBaseContracts", getContainerIds(_node.annotation().linearizedBaseContracts));
```

Comment:
What do you think about adding a helper like `makePairIfHasValue()` that would accept `SetOnce<T>` or `optional<T>` and create a pair with `null` if there's no value? `nulls` are supposed to be automatically removed now so that would let us keep the more readable compact code structure here.

Reply:
While we have the code to auto-remove nulls, I think it's better if new code doesn't even add them in the first place.
But we could extend your idea and have a `addIfSet(attributes, name, variable)`

Reply:
Right. Actually I prefer not adding nulls too but I suggested it anyway because these hairy `if`s bother me more ;) `addIfSet()` might be a good compromise.

Reply:
I replaced it now